django/db/backends/base/schema.py

```diff
@@ -1582,12 +1582,20 @@ def create_index_name(*args, **kwargs):
         )
 
     def _delete_index_sql(self, model, name, sql=None):
-        return Statement(
+        stmt = Statement(
             sql or self.sql_delete_index,
             table=Table(model._meta.db_table, self.quote_name),
             name=self.quote_name(name),
         )
 
+        # Remove all deferred statements referencing the deleted index.
+        table = stmt.parts["table"].table
+        for sql in list(self.deferred_sql):
+            if isinstance(sql, Statement) and sql.references_index(table, stmt):
+                self.deferred_sql.remove(sql)
```

Comment:
```suggestion
        for sql in list(self.deferred_sql):
            self.deferred_sql.remove(sql)
```
If I update to this, I don't see any tests failing. Can you add a test making sure not all the deferred_sql is removed when an index is deleted?